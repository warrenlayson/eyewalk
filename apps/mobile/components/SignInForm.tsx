import { yupResolver } from '@hookform/resolvers/yup'
import React from 'react'
import { useForm } from 'react-hook-form'
import { StyleSheet, TextInput, View } from 'react-native'
import * as yup from 'yup'
import Button from './Button'
import Input from './Input'
import Link from './Link'

type SignInFormProps = {
  onSignIn: (a: FormData) => void
  onForgotPassword: () => void
}

type FormData = {
  email: string
  password: string
  firstName: string
  lastName: string
}

const schema = yup.object({
  firstName: yup.string().required(),
  lastName: yup.string().required(),
  email: yup.string().email().required(),
  password: yup.string().min(6).required(),
})

const SignInForm = ({ onSignIn, onForgotPassword }: SignInFormProps) => {
  const {
    control,
    handleSubmit,
    formState: { errors, isSubmitting },
  } = useForm<FormData>({
    resolver: yupResolver(schema),
  })

  const lastNameRef = React.useRef<TextInput | null>(null)
  const emailRef = React.useRef<TextInput | null>(null)
  const passwordRef = React.useRef<TextInput | null>(null)
  const onSubmit = handleSubmit(data => {
    onSignIn(data)
  })
  return (
    <View>
      <Input
        label="First Name"
        name="firstName"
        control={control}
        error={errors.firstName?.message}
        autoCorrect={false}
        returnKeyType="next"
        blurOnSubmit={false}
        onSubmitEditing={() => lastNameRef.current?.focus()}
      />
      <Input
        label="Last Name"
        name="lastName"
        ref={lastNameRef}
        control={control}
        error={errors.lastName?.message}
        autoCorrect={false}
        blurOnSubmit={false}
        returnKeyType="next"
        onSubmitEditing={() => emailRef.current?.focus()}
      />
      <Input
        label="Email"
        control={control}
        name="email"
        ref={emailRef}
        error={errors.email?.message}
        keyboardType="email-address"
        autoCapitalize="none"
        autoCorrect={false}
        returnKeyType="next"
        onSubmitEditing={() => passwordRef.current?.focus()}
        blurOnSubmit={false}
      />
      <Input
        label="Password"
        control={control}
        name="password"
        secureTextEntry
        error={errors.password?.message}
        autoCapitalize="none"
        autoCorrect={false}
        ref={passwordRef}
        onSubmitEditing={onSubmit}
      />

      <View style={{ marginVertical: 16 }}>
        <Link onPress={onForgotPassword}>Forgot your Password?</Link>
      </View>

      <Button title="Login" onPress={onSubmit} disabled={isSubmitting} />
    </View>
  )
}

export default SignInForm

const styles = StyleSheet.create({})
