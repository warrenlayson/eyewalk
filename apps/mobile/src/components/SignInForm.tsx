import { yupResolver } from '@hookform/resolvers/yup'
import type { LoginType } from 'api/src/routes/auth/login'
import React from 'react'
import { useForm } from 'react-hook-form'
import { StyleSheet, TextInput, View } from 'react-native'
import * as yup from 'yup'
import Button from './Button'
import Input from './Input'
import Link from './Link'

type SignInFormProps = {
  onSignIn: (a: LoginType) => void
  onForgotPassword: () => void
}

const schema = yup.object({
  email: yup.string().email().required(),
  password: yup.string().min(6).required(),
})

const SignInForm = ({ onSignIn, onForgotPassword }: SignInFormProps) => {
  const {
    control,
    handleSubmit,
    formState: { errors, isSubmitting },
  } = useForm<LoginType>({
    resolver: yupResolver(schema),
    defaultValues: {
      email: '',
      password: '',
    },
  })

  const passwordRef = React.useRef<TextInput | null>(null)
  const onSubmit = handleSubmit(data => {
    onSignIn({ email: data.email.trim(), password: data.password.trim() })
  })
  return (
    <View>
      <Input
        label="Email"
        control={control}
        name="email"
        error={errors.email?.message}
        keyboardType="email-address"
        autoCapitalize="none"
        autoCorrect={false}
        returnKeyType="next"
        onSubmitEditing={() => passwordRef.current?.focus()}
        blurOnSubmit={false}
      />
      <Input
        label="Password"
        control={control}
        name="password"
        secureTextEntry
        error={errors.password?.message}
        autoCapitalize="none"
        autoCorrect={false}
        ref={passwordRef}
        onSubmitEditing={onSubmit}
      />

      <View style={{ marginVertical: 16 }}>
        <Link onPress={onForgotPassword}>Forgot your Password?</Link>
      </View>

      <Button title="Login" onPress={onSubmit} disabled={isSubmitting} />
    </View>
  )
}

export default SignInForm

const styles = StyleSheet.create({})
