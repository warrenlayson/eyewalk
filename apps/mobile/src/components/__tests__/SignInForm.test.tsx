import renderer from 'react-test-renderer'
import SignInForm from '../SignInForm'
describe('Sign in form', () => {
  test('should render correctly', async () => {
    const onSignIn = jest.fn()
    const onForgotPassword = jest.fn()
    const tree = renderer.create(
      <SignInForm onSignIn={onSignIn} onForgotPassword={onForgotPassword} />,
    )

    expect(tree).toMatchSnapshot()
  })
})
