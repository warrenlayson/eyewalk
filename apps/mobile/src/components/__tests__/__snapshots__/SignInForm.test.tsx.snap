// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sign in form should render correctly 1`] = `
<View>
  <View>
    <Text
      style={
        Object {
          "margin": 8,
          "marginLeft": 0,
        }
      }
    >
      Email
    </Text>
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      autoCorrect={false}
      blurOnSubmit={false}
      keyboardType="email-address"
      onBlur={[Function]}
      onChangeText={[Function]}
      onSubmitEditing={[Function]}
      rejectResponderTermination={true}
      returnKeyType="next"
      style={
        Array [
          Object {
            "backgroundColor": "#f7f7f7",
            "borderColor": "black",
            "borderRadius": 4,
            "height": 50,
            "padding": 16,
          },
          undefined,
        ]
      }
      underlineColorAndroid="transparent"
    />
  </View>
  <View>
    <Text
      style={
        Object {
          "margin": 8,
          "marginLeft": 0,
        }
      }
    >
      Password
    </Text>
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      autoCorrect={false}
      onBlur={[Function]}
      onChangeText={[Function]}
      onSubmitEditing={[Function]}
      rejectResponderTermination={true}
      secureTextEntry={true}
      style={
        Array [
          Object {
            "backgroundColor": "#f7f7f7",
            "borderColor": "black",
            "borderRadius": 4,
            "height": 50,
            "padding": 16,
          },
          undefined,
        ]
      }
      underlineColorAndroid="transparent"
    />
  </View>
  <View
    style={
      Object {
        "marginVertical": 16,
      }
    }
  >
    <Text
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "color": "#121212",
          },
          Object {
            "textAlign": "center",
          },
        ]
      }
    >
      Forgot your Password?
    </Text>
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#fff",
          },
          Object {
            "backgroundColor": "#ffde59",
            "borderRadius": 4,
            "fontWeight": "500",
            "height": 50,
            "justifyContent": "center",
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#121212",
            },
            Array [
              Object {
                "color": "#121212",
                "textAlign": "center",
              },
            ],
          ]
        }
      >
        Login
      </Text>
    </View>
  </View>
</View>
`;
